### ISSUE TITLE
Add a way to perform validation at the Field level

### ISSUE THREAD

**Issue by gaelduplessix** [erikras/redux-form#2185]
Hey ! First, thanks for the awesome work, I just discovered this project and it looks great.

TL;DR: It would be great to have validation at the `Field` level and not only at the `Form` level

There's a feature that I think is missing, or I didn't find any way to accomplish what I'd like to do: basically, I'd like to write custom input fields that can handle validation on their own. For example, if I create a "ColorPicker" component, I'd like it to automatically set it's error message to "Invalid color" if the user enters an invalid color.

Basically, the only needed feature would be an additional property in the `meta` object given to `Field` customComponent, that would allow it to set its validity. So in addition to `input.onChange`, `input.onBlur`, ... We could have either:
- `meta.setValidity`: a `connect`-ed action creator that would set the validity / error message for the current field
- `meta.onValidate`: or a validation method automatically called when the value is changed. Just like the `validate` property on `reduxForm` but at Field level

An other example of what it could provide, would be to write a form like this:
```jsx
<MyForm>
  <MyTextField label="First name" name="firstName" required /> // <--- "required" is handled by TextField validation
  <MyTextField label="Country" name="country" validate={(value) => isValidCountry(value)} /> // <--- Custom validation at field level =)
</MyForm>
```

I think the current validation system is very limited and does not really take advantage of redux. If I wanted to add this feature with the current API I would need to wrap my form around an other HOC which would allow fields to register via context and automatically generate the validation function at form level. Or I could create a plugin to alter the state myself but I don't think that's a good solution.

**Comment by oyeanuj** [comment]
@gaelduplessix I think you might be interested in the solution provided in #980. It essentially provides you a way to use asyncValidate and also passes which field triggered it. You can then check the `meta.valid` flag for that for your use-case.

For a more conceptual enhancement, join the discussion on #2169.

Hope that helps!

**Comment by davidkpiano** [comment]
Similar to this? http://davidkpiano.github.io/react-redux-form/docs/guides/validation.html#validation-with-field-and-control-components (adding for potential inspiration, would love to see this in Redux-Form!)

**Comment by oyeanuj** [comment]
@davidkpiano Does RRF also support something similar for submitting at a field level?

**Comment by gaelduplessix** [comment]
@davidkpiano Exactly like this ! I can't find any way to implement this with the current API. I really think validation should be responsibility of the *fields* and not the form, or at least both options should be provided.
I don't like the idea of having to write a `validate` function that checks all my required fields for every form, when just a `require` on my input component would be enough.

**Comment by alvelig** [comment]
@gaelduplessix certainly a convenience method would be great. I found the following way in the Field Component class to put `required` prop to check if it's empty:

```
<Field name="names" component={FormInput} label="Names" required />
```

FormInput class:
```
import React from 'react';

import { stopSubmit } from '/app/redux/redux-form';

export default class FormInput extends React.Component {
    componentWillUpdate(nextProps){
        const { input, name, meta, validate, required, requiredMessage } = nextProps;
        let errors = {};

        if(meta && !meta.touched) {
            return;
        }

        if(required && meta.touched && input.value == "" ) {
            errors[name] = requiredMessage || 'Este campo es obligatorio';
            stopSubmit(this, errors);
        }
    }

    render() {
        const { input, label, placeholder, containerClass, meta, ...props } = this.props;

        const { touched, error, valid } = meta || {};

        const hasError = touched && error && "has-error";
        const hasSuccess = touched && valid && "has-success";

        return (
            <div className={ [ "form-group", hasError, hasSuccess ].join(" ") }>

                <label className="col-xs-2 control-label">
                    { label }
                </label>

                <div className={containerClass || "col-sm-10"}>

                    <input {...input}
                           type="text"
                           className="form-control"
                           placeholder={ placeholder || label || "" }
                           />

                    <span className="help-block" >
                        { touched && error }
                    </span>

                </div>
            </div>
        );
    }
}
```

redux-form.js:
```
import { stopSubmit as _stopSubmit } from 'redux-form';

export const stopSubmit = (obj, errors) => {

    const { form } = obj._reactInternalInstance && obj._reactInternalInstance._context && obj._reactInternalInstance._context._reduxForm || {};

    obj.props.meta.dispatch(_stopSubmit(form, errors));

};
```

The only disadvantage that I see is that I have to access context in _reactInternalInstance. Would be great if ConnectedField passed context to the component.

**Comment by dtipson** [comment]
One option: if the validation/warn functions received an additional argument of prop data for the fields, we could compare things like "required" against the values directly.

Another would be a special case for pattern/required that hooks into validation directly, or that also creates a second argument that lists the fields that don't match their prop required/pattern so that the validation function can handle creating messages for them.  

I'll look into the code and see how complex either option might be.

**Comment by iambanh2** [comment]
It need for me. Voted :)

**Comment by erikras** [comment]
Here you go!

## Field-level validation in [`v6.3.0`](https://github.com/erikras/redux-form/releases/tag/v6.3.0)

**Comment by gaelduplessix** [comment]
Wow that's awesome ! Great job, thanks ! 👍

**Comment by levbrie** [comment]
Works great!  If we want to notify the user that there are errors in the form how can we check in the form to see if there are any field-level validation errors? I'm thinking of the case where the form is long and we displays errors for a given field but these are out of sight when the form is submitted.

**Comment by gustavohenke** [comment]
@levbrie use [`invalid` prop](http://redux-form.com/6.5.0/docs/api/Props.md/#-invalid-boolean-) at the form level, or [one of the `getForm*Errors` selectors](http://redux-form.com/6.5.0/docs/api/Selectors.md/).

**Comment by levbrie** [comment]
`syncErrors` and `submitFailed` get properly set in state here but `getFormSyncErrors` and `getFormAsyncErrors` don't seem to do anything, and getFormSubmitErrors seems to return undefined (when only using field-level validation).

**Comment by levbrie** [comment]
although using `submitFailed` and `invalid` seems to do the job here.

**Comment by gustavohenke** [comment]
> `getFormSyncErrors` and `getFormAsyncErrors` don't seem to do anything

If you could open a new issue demonstrating the problem, it would be great.

**Comment by trongphan** [comment]
If I want to configure my validation in JSON for a field like: userNameValidatorConfig = [required, {MaxLength: 50}].
I want to value 50 to be configurable rather than a fix number in function.

Thanks,
